### PR TITLE
Implement PerformanceDisparityReport

### PR DESCRIPTION
**Pull request by OlivierBinette**
#### Reference Issues/PRs

Implements PerformanceDisparityReport (see #2208).

#### Any other comments?

See https://github.com/OlivierBinette/PerformanceDisparityReport-test/blob/main/test_plots.ipynb for plot tests

A few potential features have not yet been implemented:
1. Alternative choice of baseline score (best possible score or best subgroup score).
2. Alternative ranking by *relative* score difference
3. Allowing lists of protected_feature and control_feature
4. Allowing multiple scorers (there could be a tab for each scorer).

**Note:** I wasn't able to run `make test` or `make tox` locally (I get the error "Could not find a version that satisfies the requirement torch==1.10.2+cpu"). I'm opening the PR to trigger CI/CD and there will probable be a few issues to fix with the tests.

### CODE REVIEW

**Review by noamzbr — COMMENTED** [review]
Great work! Left some comments below.

About the plot - @ptanor commented that it would be helpful to have a legend, making it clear that the black point is the group avg, and the white one is the specific score

Some comments about CI - you can run only pylint and codestyle using `make pylint` and `make docstrings`. Also, you can manually add pytorch to the env created by `make test` and then try re-running, as a hack for now.

**Comment by noamzbr** [comment]
Regarding the notebook, looks good, but there are a lot of zero division warnings, which is weird given that we set `zero_division=0` in our scorers. Do you have any idea why this happens?

**Comment by OlivierBinette** [comment]
@noamzbr Not sure why the zero division warnings happen, but I'll investigate after I've addressed the other comments. There's also issue #2214 which raises an error in the last plot test.

File: deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+from plotly.subplots import make_subplots
+
+import pandas as pd
+import numpy as np
+import itertools
+
+
+class PerformanceDisparityReport(SingleDatasetCheck):
+    """
+    Check for performance disparities across subgroups, while optionally accounting for a control variable.
```
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py:20]: Good point! I added "It is intended to be used for fairness analyses." at the end of the second paragraph.

File: deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+
+    The check identifies "performance disparities": large performance difference for a subgroup compared to the baseline performance for the full population.
+
+    Subgroups are defined by the levels of a "protected" feature. Numerical features are first binned into quantiles, while categorical features are preserved as-is. The baseline score is the overall score when all subgroups are combined. You can add conditions to flag performance differences outside of given bounds. A visual display is also provided to identify subgroups with the largest performance disparities.
```
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py]: I replaced "levels of a feature" by "categories of a feature" throughout the docstrings and documentation example.

I also replaced the variables "max_subgroups_per_category_to_display" by "max_subgroups_per_control_cat_to_display" and "max_categories_to_display" by  "max_control_cat_to_display"

File: deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+
+    Parameters
+    ----------
+    protected_feature : Hashable
+        Feature evaluated for performance disparities. Numerical features are binned into `max_segments` quantiles. Categorical features are not transformed.
+    control_feature : Hashable, default: None
+        Feature used to group data prior to evaluating performance disparities (disparities are only assessed within the groups defined by this feature). Numerical features are binned into `max_segments` quantiles. Categorical features are re-grouped into at most `max_segments` groups if necessary.
+    scorer : str, default: None
+        Name of the performance score function to use.
+    max_segments : int, default: 10
```
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py:47]: Done and I adjusted the documentation accordingly.

File: deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+        Maximum number of segments into which `control_feature` is binned.
+    min_subgroup_size : int, default: 5
```
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py:49]: Good catching this. Now subgroups where `len(data) < self.min_subgroup_size` will receive a NaN score. These scores are not displayed.

File: deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+
+class PerformanceDisparityReport(SingleDatasetCheck):
+    """
+    Check for performance disparities across subgroups, while optionally accounting for a control variable.
+
+    The check identifies "performance disparities": large performance difference for a subgroup compared to the baseline performance for the full population.
+
+    Subgroups are defined by the levels of a "protected" feature. Numerical features are first binned into quantiles, while categorical features are preserved as-is. The baseline score is the overall score when all subgroups are combined. You can add conditions to flag performance differences outside of given bounds. A visual display is also provided to identify subgroups with the largest performance disparities.
+
+    Additionally, the analysis may be separated across the levels of a "control" feature. Numerical features are binned and categorical features are re-binned into `max_number` categories. To account for the control feature, baseline scores and subgroup scores are be computed within each of its levels.
```
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py]: I added the following paragraph:

"The check's display is a plot of the baseline score (black point) and subgroup scores (white point)
    for each subgroup, faceted by control feature categories, and sorted by the size of the 
    performance disparity. The performance disparity is highlighted by a red line for negative differences
    and a green line for positive differences. By default, only the top 3 categories of the control 
    feature with the largest performance disparities are displayed. Within each control category 
    subplot, the top 3 subgroups with the largest performance disparities are displayed."

File: deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+        Name of the performance score function to use.
+    max_segments : int, default: 10
+        Maximum number of segments into which `control_feature` is binned.
+    min_subgroup_size : int, default: 5
+        Minimum size of a subgroup for which to compute a performance score.
+    max_subgroups_per_category_to_display : int, default: 3
+        Maximum number of subgroups to display.
+    max_categories_to_display: int, default: 3
+        Maximum number of `control_feature` categories to display.
+    use_avg_defaults : bool, default: True
```
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py:55]: Removed it.

File: deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+
+    Additionally, the analysis may be separated across the levels of a "control" feature. Numerical features are binned and categorical features are re-binned into `max_number` categories. To account for the control feature, baseline scores and subgroup scores are be computed within each of its levels.
+
+    Parameters
+    ----------
+    protected_feature : Hashable
+        Feature evaluated for performance disparities. Numerical features are binned into `max_segments` quantiles. Categorical features are not transformed.
+    control_feature : Hashable, default: None
+        Feature used to group data prior to evaluating performance disparities (disparities are only assessed within the groups defined by this feature). Numerical features are binned into `max_segments` quantiles. Categorical features are re-grouped into at most `max_segments` groups if necessary.
+    scorer : str, default: None
```
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py]: I updated it so that it has the following signature:

```
scorer : str or Tuple[str, Union[str, Callable]], default: None
```

Code to get the scorer is:
```python
if isinstance(self.scorer, str):
    scorer = context.get_single_scorer({self.scorer: self.scorer})
else:
    scorer = context.get_single_scorer(dict([self.scorer]))
```

It is convenient to be able to pass a single string as well. What do you think?

File: deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+            if len(data) == 0:
+                if classwise:
+                    return {cls: np.nan for cls in model_classes}
+                else:
+                    return np.nan
+            return scorer(model, dataset.copy(data))
+        def apply_scorer(x):
+            return score(x["_dataset"], model, x["_scorer"])
+
+        scores_df["_score"] = scores_df.apply(apply_scorer, axis=1)
```
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py:204]: Good point, added comments.

File: deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+        if self.protected_feature not in data.columns:
+            raise DeepchecksValueError(f"Feature {self.protected_feature} not found in dataset.")
+
+        if self.control_feature is not None and self.control_feature not in data.columns:
+            raise DeepchecksValueError(f"Feature {self.control_feature} not found in dataset.")
+
+        if self.control_feature is not None and self.control_feature == self.protected_feature:
+            raise DeepchecksValueError(f"protected_feature {self.control_feature} and control_feature {self.protected_feature} are the same.")
+
+    def make_partitions(self, dataset):
```
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py:163]: Good point, made the change for all private methods.

File: deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+        # Get distinct subplot categories with the largest observed differences
+        if len(subplot_grouping) > 0:
+            subplots_categories = (
+                visual_df.sort_values("_diff", ascending=True)[subplot_grouping]
+                .drop_duplicates()
+                .head(self.max_categories_to_display)
+            )
+            rows = len(subplots_categories)
+            if rows == 0:
+                raise DeepchecksProcessError("Nothing to display.")
```
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py:307]: This happens when you are grouping by control_feature or model_class, but the subgroups are too small to have any computed scores. I replaced the error with the following:

```python
if len(visual_df) == 0:
    # Return text figure: "No scores to display"
    fig = go.Figure()
    fig.update_layout(xaxis_visible=False, yaxis_visible=False)
    fig.add_annotation(
        text="No scores to display",
        xref="paper",
        yref="paper",
        showarrow=False,
        font=dict(size=28),
    )
    return fig
```

[the remark this answers is not in the document]
Comment (OlivierBinette) [reply deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py:307]: This is what it looks like: 
screenshot

File: docs/source/checks/tabular/model_evaluation/plot_performance_disparity_report.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+* `Generate data & model <#generate-data-model>`__
+* `Run the check <#run-the-check>`__
+* `Define a condition <#define-a-condition>`__
+
+What is the purpose of the check?
+==================================
+
+The check is designed to help you identify subgroups for which the model has a much lower performance score than its baseline score (its overall performance). The subgroups are defined by a chosen *protected* feature (e.g., "sex", "race") and you can specify a *control* feature (e.g., "education") by which to group the data before computing performance differences.
+
+Large performance disparities can indicate a problem with the model. The training data may not be sufficient for certain subgroups or may contain biases, or the model may need to be re-calibrated when applied to certain subgroups. When using appropriate scoring functions, looking at performance disparities can help uncover issues of these kinds.
```
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply docs/source/checks/tabular/model_evaluation/plot_performance_disparity_report.py:21]: Added "It is intended to be used for fairness analyses." in the second paragraph of the class documentation

File: docs/source/checks/tabular/model_evaluation/plot_performance_disparity_report.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+Performance Disparity Report
+*************************
+
+This notebook provides an overview for using and understanding the Performance Disparity Report check.
+
+**Structure:**
+
+* `What is the purpose of the check? <#what-is-the-purpose-of-the-check>`__
+* `Automatically detecting weak segments <#automatically-detecting-weak-segments>`__
+* `Generate data & model <#generate-data-model>`__
```
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply docs/source/checks/tabular/model_evaluation/plot_performance_disparity_report.py:12]: @noamzbr I'm not too sure which template to follow here. Can you refer to an existing example that I can use as a template?

File: docs/source/checks/tabular/model_evaluation/plot_performance_disparity_report.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+"""
+Performance Disparity Report
+*************************
+
+This notebook provides an overview for using and understanding the Performance Disparity Report check.
+
+**Structure:**
+
+* `What is the purpose of the check? <#what-is-the-purpose-of-the-check>`__
+* `Automatically detecting weak segments <#automatically-detecting-weak-segments>`__
```
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply docs/source/checks/tabular/model_evaluation/plot_performance_disparity_report.py:11]: Fixed

File: docs/source/checks/tabular/model_evaluation/plot_performance_disparity_report.py [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+#%%
+# Run the check
+# =============
+#
+# The check requires the argument ``protected_feature`` identifying a column that defines the subgroups
+# for which performance disparities are assessed. In addition, the check has several optional parameters
+# that affect its behavior and output.
+# 
+# ``control_feature``: Column to use to split the data by groups prior to computing performance disparities.
```
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply docs/source/checks/tabular/model_evaluation/plot_performance_disparity_report.py:44]: Done.

File: docs/source/checks/tabular/model_evaluation/plot_performance_disparity_report.py [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+# 
+# ``scorer``: Scoring function to measure performance.
```
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply docs/source/checks/tabular/model_evaluation/plot_performance_disparity_report.py:46]: Added: "Default to "accuracy" for classification tasks and "r2" for regression tasks."

File: tests/tabular/checks/model_evaluation/performance_disparity_report_test.py [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+import numpy as np
+import pandas as pd
+from hamcrest import assert_that, close_to, equal_to, has_items, has_length, calling, raises, any_of
+from sklearn.metrics import f1_score, make_scorer
+
+from deepchecks.core.errors import DeepchecksValueError, DeepchecksNotSupportedError
+from deepchecks.tabular.checks.model_evaluation import PerformanceDisparityReport
+from tests.base.utils import equal_condition_result
+
+def test_no_error(adult_split_dataset_and_model, avocado_split_dataset_and_model):
```
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply tests/tabular/checks/model_evaluation/performance_disparity_report_test.py:23]: Added tests for numeric values.

**Comment by OlivierBinette** [comment]
I added a legend to the plots. It can be clicked to hide/show positive/negative differences:
screenshot

**Comment by noamzbr** [comment]
Closes #2208

File: docs/source/checks/tabular/model_evaluation/plot_performance_disparity_report.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply docs/source/checks/tabular/model_evaluation/plot_performance_disparity_report.py:11]: You are right that this is the structure for our other docs, so let's keep it that way for now.

File: docs/source/checks/tabular/model_evaluation/plot_performance_disparity_report.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply docs/source/checks/tabular/model_evaluation/plot_performance_disparity_report.py:20]: I think it should appear here as well (I expect more people will read this rather than the class docs)

**Review by noamzbr — APPROVED** [review]
Approved, two minor comments remaining

File: deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py]: Good catch, we should sort it later. For now I think what you did makes sense (most people will want to just pass a string)

File: deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py]: 1. You can just have text rather than a plotly graph 
2. Perhaps tell the user that we're not displaying anything because there are not enough samples.

File: docs/source/checks/tabular/model_evaluation/plot_performance_disparity_report.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply docs/source/checks/tabular/model_evaluation/plot_performance_disparity_report.py:20]: Added the following to the documentation example:

"This is primarily useful for fairness analyses, but can also be used to identify other types of performance disparities."

File: deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (OlivierBinette) [reply deepchecks/tabular/checks/model_evaluation/performance_disparity_report.py]: Replaced the plotly Figure with a text string: f"No scores to display. Subgroups may be smaller than the minimum size of {self.min_subgroup_size}."

**Comment by OlivierBinette** [comment]
**Important note:** I added a requirement for pandas>=1.3, which is required for my use of pandas' "explode()" function with multiple columns. However, pandas 1.3 does not support Python 3.6, so this would remove Python 3.6 compatibility.

So I think we need to change the code to be compatible with pandas 1.1.5 and Python 3.6, and revert the requirements change.

**Comment by OlivierBinette** [comment]
> Regarding the notebook, looks good, but there are a lot of zero division warnings, which is weird given that we set `zero_division=0` in our scorers. Do you have any idea why this happens?

This is because I used the "f1" scorer which is not defined by Deepchecks, but rather refers to sklearn's metric directly. Deepchecks defines "f1_macro", "f1_micro", "f1_weighted", "f1_per_class".

**Comment by noamzbr** [comment]
Thanks @OlivierBinette. Regarding the .explode - agree that it should be replaced with something more backwards-compatible.
Regarding the f1_score - understood.

If you don't have the time right now to work on this (Merry Christmas by the way!), you can give me permissions to work on your branch and we can take it from here, or alternatively you can open a new PR into the deepchecks:performance-disparity branch and then we can work on it ourselves.

**Comment by OlivierBinette** [comment]
@noamzbr I opened #2225. Thanks!

**Comment by noamzbr** [comment]
Will be merged in https://github.com/deepchecks/deepchecks/pull/2226
